File: leetcode_problems/typescript/src/lc844_backspace_string_compare.ts
function toStackFromString(s: string): string[] {
  let st: string[] = [];
  for (let chr of s) {
    if (chr == "#") {
      if (st.length > 0) {
        st.pop();
      }
    } else {
      st.push(chr);
    }
  }
  return st;
}

function backspaceCompare(s: string, t: string): boolean {
  let st1 = toStackFromString(s);
  let st2 = toStackFromString(t);
  if (st1.length != st2.length) {
    return false;
  }
  while (st1.length > 0) {
    if (st1[st1.length - 1] != st2[st2.length - 1]) {
      return false;
    }
    st1.pop();
    st2.pop();
  }
  return true;
}

File: leetcode_problems/cpp/lc844_backspace_string_compare.cpp
#include <string>
#include <stack>
using std::stack;
using std::string;

class Solution
{
public:
    stack<char> to_stack(string s)
    {
        stack<char> st;
        for (char chr : s)
        {
            if (chr == '#')
            {
                if (!st.empty())
                {
                    st.pop();
                }
            }
            else
            {
                st.push(chr);
            }
        }
        return st;
    }

    bool backspaceCompare(string s, string t)
    {
        stack<char> st1 = to_stack(s);
        stack<char> st2 = to_stack(t);
        if (st1.size() != st2.size())
        {
            return false;
        }
        while (!st1.empty())
        {
            if (st1.top() != st2.top())
            {
                return false;
            }
            st1.pop();
            st2.pop();
        }
        return true;
    }
};